api/buckets/service.go

```diff
@@ -1584,6 +1809,155 @@ func (s *Service) removeNodeAtPath(ctx context.Context, pth path.Path, key []byt
 	return path.IpfsPath(np[0].new.Cid()), nil
 }
 
+func (s *Service) PushPathAccessRoles(ctx context.Context, req *pb.PushPathAccessRolesRequest) (*pb.PushPathAccessRolesResponse, error) {
+	log.Debugf("received push path access roles request")
+
+	dbID, ok := common.ThreadIDFromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("db required")
+	}
+	dbToken, _ := thread.TokenFromContext(ctx)
+
+	reqPath, err := parsePath(req.Path)
+	if err != nil {
+		return nil, err
+	}
+	buck, pth, err := s.getBucketPath(ctx, dbID, req.Key, reqPath, dbToken)
+	if err != nil {
+		return nil, err
+	}
+	linkKey := buck.GetLinkEncryptionKey()
+	pathNode, err := s.getNodeAtPath(ctx, pth, linkKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var currentFileKeys map[string][]byte
+	md, mdPath, ok := buck.GetMetadataForPath(reqPath, false)
+	if !ok {
+		return nil, fmt.Errorf("could not resolve path: %s", reqPath)
+	}
+	var target tdb.Metadata
+	if mdPath != reqPath { // If the metadata is inherited from a parent, create a new entry
+		target = tdb.Metadata{
+			Roles: make(map[string]buckets.Role),
+		}
+	} else {
+		target = md
+	}
+	if buck.IsPrivate() {
+		currentFileKeys, err = buck.GetFileEncryptionKeysForPrefix(reqPath)
+		if err != nil {
+			return nil, err
+		}
+		newFileKey, err := dcrypto.NewKey()
+		if err != nil {
+			return nil, err
+		}
+		target.SetFileEncryptionKey(newFileKey) // Create or refresh the file key
+	}
+
+	newRoles, err := buckets.RolesFromPb(req.Roles)
+	if err != nil {
+		return nil, err
+	}
+	var changed bool
+	for k, r := range newRoles {
+		if x, ok := target.Roles[k]; ok && x == r {
+			continue
+		}
+		if r > 0 {
+			target.Roles[k] = r
+		} else {
+			delete(target.Roles, k)
+		}
+		changed = true
+	}
+	if changed {
```

Comment:
This method edits the access roles for the requested path. The idea here is to determine if the request, which contains a map of identities and roles, actually amounts to a change.